crates/flowistry/benches/main.rs

```diff
@@ -0,0 +1,201 @@
+  ty::TyCtxt,
+};
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum AnalysisType {
+  FlowOnly,
+  FlowAndDeps,
+}
+
+fn analysis<'tcx>(
+  tcx: TyCtxt<'tcx>,
+  body_id: BodyId,
+  body_with_facts: &BodyWithBorrowckFacts<'tcx>,
+  ty: AnalysisType,
+) {
+  let results = flowistry::infoflow::compute_flow(tcx, body_id, body_with_facts);
+
+  if ty == AnalysisType::FlowAndDeps {
+    let targets = body_with_facts
+      .body
+      .local_decls
+      .indices()
+      .map(|local| {
+        let arg = Place::make(local, &[], tcx);
+        vec![(arg, Location::START)]
+      })
+      .collect::<Vec<_>>();
+
+    flowistry::infoflow::compute_dependencies(&results, targets, Direction::Both);
+  }
+}
+
+struct UnsafeBenchGroup(
+  criterion::BenchmarkGroup<'static, criterion::measurement::WallTime>,
+);
+
+impl UnsafeBenchGroup {
+  fn new(bench_group: BenchmarkGroup<WallTime>) -> Self {
+    unsafe {
+      UnsafeBenchGroup(std::mem::transmute::<
+        criterion::BenchmarkGroup<'_, criterion::measurement::WallTime>,
+        criterion::BenchmarkGroup<'static, criterion::measurement::WallTime>,
+      >(bench_group))
+    }
+  }
+}
+
+// SAFETY: Rustc requires that Callbacks implements Send in case the compiler is
+// run with multiple threads (which we don't do)
+unsafe impl Send for UnsafeBenchGroup {}
+
+struct Callbacks {
+  group: UnsafeBenchGroup,
+}
+impl rustc_driver::Callbacks for Callbacks {
+  fn config(&mut self, config: &mut rustc_interface::Config) {
+    config.override_queries = Some(borrowck_facts::override_queries);
+  }
+
+  fn after_parsing<'tcx>(
+    &mut self,
+    _compiler: &rustc_interface::interface::Compiler,
+    queries: &'tcx rustc_interface::Queries<'tcx>,
+  ) -> rustc_driver::Compilation {
+    queries.global_ctxt().unwrap().take().enter(|tcx| {
+      let hir = tcx.hir();
+      let body_id = hir
+        .items()
+        .filter_map(|item| match item.kind {
+          ItemKind::Fn(_, _, body) => Some(body),
+          _ => None,
+        })
+        .next()
+        .unwrap();
+
+      let def_id = hir.body_owner_def_id(body_id);
+      let body_with_facts = borrowck_facts::get_body_with_borrowck_facts(tcx, def_id);
+
+      for analysis_ty in [AnalysisType::FlowOnly, AnalysisType::FlowAndDeps] {
+        let bench_id = match analysis_ty {
+          AnalysisType::FlowOnly => "Flow",
+          AnalysisType::FlowAndDeps => "Flow + Deps",
+        };
+
+        self.group.0.bench_function(bench_id, |b| {
+          b.iter(|| analysis(tcx, body_id, body_with_facts, analysis_ty))
+        });
+      }
+    });
+    rustc_driver::Compilation::Stop
+  }
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+  const TESTS: &[(&str, &str)] = &[
+    ("Locations", "locations.rs"),
+    ("Unique Lifetimes", "lifetimes_unique.rs"),
+    ("Infoflow", "infoflow.rs"),
+    ("Places", "places.rs"),
+    ("Same Lifetime", "lifetimes_same.rs"),
+    ("Nested Structs", "nested_struct.rs"),
+  ];
+
+  (|| -> Result<()> {
```

Comment:
There are lots of unwraps() in this function that should `?` or `.with_context(..)`?

Reply:
Ah yes, should've caught that

Reply:
Also, do you have a preference between `context()` and `with_context()`? From what I understand, `with_context` should be used to lazily evaluate the message in case it's expensive to calculate. Personally `context()` looks slightly cleaner because it omits the closure but I'm fine with either.

Reply:
Your understanding is correct. So my general rule: `.context` for literals, `.with_context` for calculations. So `.context("Hey")` and `.with_context(|| format!("{name}: hey"))`. Similar rule for `.unwrap_or` vs `.unwrap_or_else`.

Reply:
That should be all of them. I kept the unwraps for functions that (hopefully) shouldn't fail like `to_str` and for `Result`s I can't use `context` with (`Result<R, ErrorReported>` from `rustc_driver::catch_fatal_errors`).